.github/workflows/rust.yml

```diff
@@ -21,145 +21,286 @@ on:
     strategy:
-      fail-fast: false
       matrix:
+        arch: [amd64]
         rust: [nightly-2020-11-24]
-    env:
-      RUST: ${{ matrix.rust }}
+    container:
+      image: ${{ matrix.arch }}/rust
     steps:
-      - name: Checkout Arrow
-        uses: actions/checkout@v2
+      - uses: actions/checkout@v2
+      - name: Cache Cargo
+        uses: actions/cache@v2
         with:
-          fetch-depth: 0
-      - name: Fetch Submodules and Tags
-        shell: bash
-        run: ci/scripts/util_checkout.sh
-      - name: Free Up Disk Space
-        shell: bash
-        run: ci/scripts/util_cleanup.sh
-      - name: Cache Docker Volumes
-        uses: actions/cache@v1
-        with:
-          path: .docker
-          key: debian-10-rust-${{ hashFiles('rust/**/**.rs', 'rust/**/Cargo.toml') }}
-          restore-keys: debian-10-rust-
-      - name: Setup Python
-        uses: actions/setup-python@v1
-        with:
-          python-version: 3.8
-      - name: Setup Archery
-        run: pip install -e dev/archery[docker]
-      - name: Execute Docker Build
+          # these represent dependencies downloaded by cargo
+          # and thus do not depend on the OS, arch nor rust version.
+          path: /github/home/.cargo
+          key: cargo-cache-
+      - name: Cache Rust dependencies
+        uses: actions/cache@v2
+        with:
+          # these represent compiled steps of both dependencies and arrow
+          # and thus are specific for a particular OS, arch and rust version.
+          path: /github/home/target
+          key: ${{ runner.os }}-${{ matrix.arch }}-target-cache-${{ matrix.rust }}-
+      - uses: actions-rs/toolchain@v1
+        with:
+          toolchain: ${{ matrix.rust }}
+          default: true
+          components: rustfmt
+      - name: Build
         run: |
-          sudo sysctl -w kernel.core_pattern="core.%e.%p"
-          ulimit -c unlimited
-          archery docker run debian-rust
-      - name: Fix Cache Permissions
-        run: sudo chmod -R o+r .docker
-      - name: Docker Push
-        if: success() && github.event_name == 'push' && github.repository == 'apache/arrow'
-        continue-on-error: true
-        shell: bash
-        run: archery docker push debian-rust
+          export CARGO_HOME="/github/home/.cargo"
+          export CARGO_TARGET_DIR="/github/home/target"
+          cd rust
+          cargo build
 
-  windows:
-    name: AMD64 Windows 2019 Rust ${{ matrix.rust }}
-    runs-on: windows-latest
-    if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
+  # test the crate
+  linux-test:
+    name: Linux ${{ matrix.arch }} Rust ${{ matrix.rust }} test workspace
+    needs: [linux-build-lib]
+    runs-on: ubuntu-latest
     strategy:
-      fail-fast: false
       matrix:
+        arch: [amd64]
         rust: [nightly-2020-11-24]
+    container:
+      image: ${{ matrix.arch }}/rust
+      env:
+        ARROW_TEST_DATA: /__w/arrow/arrow/testing/data
```

Comment:
What is the `__w` prefix?

Reply:
It is the absolute path to where the container action places the checkout code on. I tried using relative paths but I was unable to make it work.